File: algorithms/BinarySearch.cpp
#include <iostream>

int binarySearch(int arr[], int size_t, int value)
{
    int left = 0;
    int right = size_t - 1;
    while (left <= right)
    {
        int middle = (left + right) / 2;
        if (value == arr[middle])
            return middle;
        else if (value > arr[middle])
            left = middle + 1;
        else
            right = middle - 1;
    }
    return -1;
}

int main()
{
    int arr[5]{3, 2, 1, 5, 4};
    int size_t = sizeof(arr) / sizeof(arr[0]);

    std::cout << binarySearch(arr, size_t, 5); // O(log(2n))
}

File: algopedia/src/main/kotlin/algorithms/BinarySearch.kt
package algorithms

class BinarySearch {

    fun execute(arr: List<Int>, value: Int): Int{
        var start = 0
        var end = arr.size - 1

        while(start <= end){
            val middle = (start + end) / 2

            when{
                arr[middle] == value -> return middle
                arr[middle] < value -> start = middle + 1
                else -> end = middle - 1
            }
        }

        return -1
    }
}